File: handy-java/quiz/q22_daily_temperatures_medium.java
import java.util.ArrayDeque;
import java.util.Arrays;

/**
 * Q22) Daily Temperatures
 *
 * AI-BEST:
 * Monotonic decreasing stack of indices.
 * Time: O(n), Space: O(n)
 *
 * AI-EASY:
 * For each day, scan forward for first warmer day.
 * Time: O(n^2), Space: O(1) extra
 */
public class q22_daily_temperatures_medium {
    static int[] dailyTemperaturesBest(int[] temperatures) {
        int n = temperatures.length;
        int[] answer = new int[n];
        ArrayDeque<Integer> stack = new ArrayDeque<>();

        // Step 1: Iterate once and resolve waiting indices when warmer day appears.
        // Why: each index is pushed/popped once, giving linear complexity.
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && temperatures[i] > temperatures[stack.peek()]) {
                int prev = stack.pop();
                answer[prev] = i - prev;
            }

            // Step 2: Push unresolved day index.
            stack.push(i);
        }

        // Step 3: Remaining stack entries already default to 0.
        return answer;
    }

    static int[] dailyTemperaturesEasy(int[] temperatures) {
        int n = temperatures.length;
        int[] answer = new int[n];

        // Step 1: For each day, scan days ahead.
        // Why: straightforward baseline, easy to understand and debug.
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (temperatures[j] > temperatures[i]) {
                    answer[i] = j - i;
                    break;
                }
            }
        }

        // Step 2: Return waits (or 0 where no warmer day exists).
        return answer;
    }

    public static void main(String[] args) {
        System.out.println("Q22: Daily Temperatures");
        int[][] cases = {
                {73, 74, 75, 71, 69, 72, 76, 73},
                {30, 40, 50, 60},
                {30, 60, 90}
        };

        for (int[] temperatures : cases) {
            System.out.println("Input: temperatures=" + Arrays.toString(temperatures));
            System.out.println("  BEST: " + Arrays.toString(dailyTemperaturesBest(temperatures)));
            System.out.println("  EASY: " + Arrays.toString(dailyTemperaturesEasy(temperatures)));
        }
    }
}

File: handy-js/quiz/q22_daily_temperatures_medium.js
/**
 * Q22) Daily Temperatures
 *
 * AI-BEST:
 * Monotonic decreasing stack of indices.
 * Time: O(n), Space: O(n)
 *
 * AI-EASY:
 * For each day, scan ahead to find first warmer day.
 * Time: O(n^2), Space: O(1) extra
 */

function dailyTemperaturesBest(temperatures) {
  const n = temperatures.length;
  const answer = new Array(n).fill(0);
  const stack = [];

  // Step 1: Walk through days once.
  // Why: stack tracks unresolved days waiting for a warmer temperature.
  for (let i = 0; i < n; i++) {
    while (
      stack.length &&
      temperatures[i] > temperatures[stack[stack.length - 1]]
    ) {
      const prevIndex = stack.pop();
      answer[prevIndex] = i - prevIndex;
    }

    // Step 2: Push current day as unresolved.
    stack.push(i);
  }

  // Step 3: Remaining indices already default to 0.
  return answer;
}

function dailyTemperaturesEasy(temperatures) {
  const n = temperatures.length;
  const answer = new Array(n).fill(0);

  // Step 1: Try each day as starting point.
  for (let i = 0; i < n; i++) {
    // Step 2: Scan forward for first warmer day.
    for (let j = i + 1; j < n; j++) {
      if (temperatures[j] > temperatures[i]) {
        answer[i] = j - i;
        break;
      }
    }
  }

  // Step 3: Return computed waits.
  return answer;
}

function runDemo() {
  console.log("Q22: Daily Temperatures");
  const cases = [
    [73, 74, 75, 71, 69, 72, 76, 73],
    [30, 40, 50, 60],
    [30, 60, 90],
  ];

  for (const temps of cases) {
    console.log(`Input: temperatures=${JSON.stringify(temps)}`);
    console.log("  BEST:", JSON.stringify(dailyTemperaturesBest(temps)));
    console.log("  EASY:", JSON.stringify(dailyTemperaturesEasy(temps)));
  }
}

runDemo();
